### PR TITLE
cleaning and enhancement to PolyDict

### LINKED ISSUE #34000: cleaning and enhancement to PolyDict
<div id="comment:0"></div>

We provide some cleaning and speed up to the `PolyDict` class by always enforcing keys to be `ETuple` (some parts of the code did assume that was the case). As a consequence we deprecate the usage of the constructor arguments `force_int_exponents` and `force_etuples`.

We also simplify the design by getting rid of `PolyDict.__zero` (and hence deprecating the argument `zero` of the `PolyDict` constructor).

Along the way, we fix

```
sage: R = ZZ['x']['y,z']
sage: y, z = R.gens()
sage: y.integral(y)
1/2*y^2
sage: parent(y.integral(y))
Multivariate Polynomial Ring in y, z over Univariate Polynomial Ring in x over Integer Ring
```
and

```
sage: from sage.rings.polynomial.polydict import ETuple
sage: e = ETuple([0, 1, 0])
sage: e.eadd_p(0, 0).nonzero_positions()
[0, 1]
```

CC:  @xcaruso @mezzarobba @saraedum

Component: **algebra**

Author: **Vincent Delecroix**

_Issue created by migration from https://trac.sagemath.org/ticket/34000_

### PR DESCRIPTION
**Pull request by videlec**
### 📚 Description

Fixes #34000 (and more complete version of the proposition at #35020).

A discussion about the general problem of inexact zeros has been opened at #35319 

### 📝 Checklist

- [x] I have made sure that the title is self-explanatory and the description concisely explains the PR.
- [x] I have linked an issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by roed314** [comment]
@videlec I'm currently in Chamonix with bad internet; I'll try to take a look at this over the weekend.

File: src/sage/rings/polynomial/polydict.pyx [hunk 1]
```diff
@@ -42,131 +43,238 @@ from cpython.dict cimport *
...
+            sage: from sage.rings.polynomial.polydict import PolyDict, monomial_exponent
+            sage: monomial_exponent(PolyDict({(2, 3): 1}))
+            (2, 3)
+            sage: monomial_exponent(PolyDict({(2, 3): 3}))
+            Traceback (most recent call last):
+            ...
+            ValueError: not a monomial
+            sage: monomial_exponent(PolyDict({(1, 0): 1, (0, 1): 1}))
+            Traceback (most recent call last):
+            :::
```
Comment (fchapoton) [inline src/sage/rings/polynomial/polydict.pyx]: should be `...`

File: src/sage/rings/polynomial/multi_polynomial_element.py [hunk 2]
```diff
@@ -333,9 +385,9 @@ def _div_(self, right):
             sage: x/S(2)
             1/2*x
         """
-        if right in self.base_ring():
-            inv = self.base_ring().one()/self.base_ring()(right)
-            return inv*self
+        if right.is_constant():
+            inv = self.base_ring().one() / self.base_ring()(right)
```
Comment (tscrim) [inline src/sage/rings/polynomial/multi_polynomial_element.py]:
```suggestion
            inv = self.base_ring().one() / right.constant_coefficient()
```
This is faster:
```python
sage: R.<x,y> = PolynomialRing(CyclotomicField(5)['q'].fraction_field())
sage: p = R(5)
sage: K = R.base_ring()
sage: %timeit p.constant_coefficient()
774 ns ± 30 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
sage: %timeit K(p)
2.14 µs ± 168 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```
Although I am not sure if `~right.constant_coefficient()` is better.

File: src/sage/rings/polynomial/multi_polynomial_ring.py [hunk 3]
```diff
@@ -891,11 +888,34 @@ def addwithcarry(tempvector, maxvector, pos):
...
+            sage: y, z = S.gens()
+            sage: S.sum([x*y, 2*x^2*z - 2*x*y, 1 + y + z])
+            (-x + 1)*y + (2*x^2 + 1)*z + 1
+
+        Comparison with builtin sum::
+
+            sage: sum([x*y, 2*x^2*z - 2*x*y, 1 + y + z])
+            (-x + 1)*y + (2*x^2 + 1)*z + 1
+        """
+        ans = self(0)
```
Comment (tscrim) [inline src/sage/rings/polynomial/multi_polynomial_ring.py]:
```suggestion
        ans = copy(self.zero())
```
or perhaps if there is a faster way to make a mutable copy. I didn't check which way is fastest.

File: src/sage/rings/polynomial/polydict.pyx [hunk 4]
```diff
@@ -42,131 +43,238 @@ from cpython.dict cimport *
...
+    """
+    if len(p.__repn) != 1 or not next(iter(p.__repn.values())).is_one():
+        raise ValueError('not a monomial')
+    return next(iter(p.__repn))
 
 
 cdef class PolyDict:
-    def __init__(PolyDict self, pdict, zero=0, remove_zero=False, force_int_exponents=True, force_etuples=True):
+    r"""
+    Datastructure for multivariate polynomials.
```
Comment (tscrim) [inline src/sage/rings/polynomial/polydict.pyx]:
```suggestion
    Data structure for multivariate polynomials.
```

File: src/sage/rings/polynomial/polydict.pyx [hunk 5]
```diff
@@ -42,131 +43,238 @@ from cpython.dict cimport *
...
-                        del pdict[k]
-        self.__repn = <dict> pdict
-        self.__zero = zero
+            raise TypeError("pdict must be a dict or a list")
+
+        for k in list(v):
+            if type(k) is not ETuple:
+                val = v[k]
+                del v[k]
+                v[ETuple(k)] = val
```
Comment (tscrim) [inline src/sage/rings/polynomial/polydict.pyx]: This is generally slower than what it was before as now it explicitly checks everything everytime for the keys being an `ETuple`. I am -1 on removing the argument that disabled this check (which we can assume for internal code that things are good).

File: src/sage/rings/polynomial/polydict.pyx [hunk 6]
```diff
@@ -315,96 +423,99 @@ cdef class PolyDict:
...
+            return -1
+        return max((<ETuple> e).get_exp(i) for e in self.__repn)
+
+    def total_degree(self, tuple w=None):
+        r"""
+        Return the total degree.
+
+        INPUT:
+
+        - ``w`` (optional) -- a tuple of weights
```
Comment (tscrim) [inline src/sage/rings/polynomial/polydict.pyx]:
```suggestion
        - ``w`` -- (optional) a tuple of weights
```

File: src/sage/rings/polynomial/polydict.pyx [hunk 7]
```diff
@@ -700,48 +842,86 @@ cdef class PolyDict:
...
-                    if not atomic_coefficients:
-                        c = str(c)
-                        if c.find("+") != -1 or c.find("-") != -1 or c.find(" ") != -1:
-                            c = "(%s)" % c
-                    multi = "%s*%s" % (c, multi)
-
-                # Now add on coefficiented multinomials
+                        multi = multi + "^(%s)" % e[j]
+            # Next determine coefficient of multinomial
+            if len(multi) == 0:
```
Comment (tscrim) [inline src/sage/rings/polynomial/polydict.pyx]:
```suggestion
            if not multi:
```

File: src/sage/rings/polynomial/polydict.pyx [hunk 8]
```diff
@@ -700,48 +842,86 @@ cdef class PolyDict:
...
+                multi = str(c)
+            elif c == neg_one and not is_characteristic_2:
+                # handle -1 specially because it's a pain
                 if len(poly) > 0:
```
Comment (tscrim) [inline src/sage/rings/polynomial/polydict.pyx]:
```suggestion
                if poly:
```
and similar.

File: src/sage/rings/polynomial/polydict.pyx [hunk 9]
```diff
@@ -926,42 +1096,113 @@ cdef class PolyDict:
...
+            PolyDict with representation {(1, 2): 0, (2, 1): 0, (2, 3): 0}
+        """
+        cdef dict D = self.__repn.copy()
+        cdef dict R = other.__repn
+        for e, c in R.items():
+            if e in D:
+                D[e] -= c
+            else:
+                D[e] = -c
+        return self._new(D)
```
Comment (tscrim) [inline src/sage/rings/polynomial/polydict.pyx:1195]: You need to also need to remove the zeros from here.

File: src/sage/rings/polynomial/polydict.pyx [hunk 10]
```diff
@@ -1171,13 +1414,29 @@ cdef class ETuple:
...
+    def __cinit__(self):
         self._data = <int*>0
 
     def __dealloc__(self):
         if self._data != <int*>0:
             sig_free(self._data)
 
+    def __bool__(self):
+        r"""
+        Return whether this polynomial is nonzero.
```
Comment (tscrim) [inline src/sage/rings/polynomial/polydict.pyx]:
```suggestion
        Return whether this ``ETuple`` is nonzero.
```
Although this is why I say `self` is better...

File: src/sage/rings/polynomial/polydict.pyx [hunk 11]
```diff
@@ -1467,68 +1726,75 @@ cdef class ETuple:
...
+             sage: ETuple([-1, 1]).unweighted_degree()
+             0
         """
-        cdef size_t degree = 0
+        cdef int degree = 0
         cdef size_t i
-        for i in range(1, 2*self._nonzero, 2):
-            degree += self._data[i]
+        for i in range(self._nonzero):
+            degree += self._data[2 * i + 1]
```
Comment (tscrim) [inline src/sage/rings/polynomial/polydict.pyx:1858]: Did you check the C code output here? I have lost track if Cython has been able to make these types of loops much more efficient into a simple `for(i=0; i<self._nonzero; i+=2)` or not. It might be good to leave a `TODO` comment here if the C code is still going through Python.

File: src/sage/rings/polynomial/polydict.pyx [hunk 12]
```diff
@@ -1467,68 +1726,75 @@ cdef class ETuple:
...
         cdef size_t i
-        cdef size_t deg = 0
-        assert len(w) == self._length
-        for i in range(0, 2*self._nonzero, 2):
-            deg += <size_t> self._data[i+1] * <size_t> w[self._data[i]]
+        cdef int deg = 0
+        if len(w) != self._length:
+            raise ValueError
+        for i in range(0, 2 * self._nonzero, 2):
+            deg += self._data[i+1] * <int> w[self._data[i]]
```
Comment (tscrim) [inline src/sage/rings/polynomial/polydict.pyx]: Essentially the same comment as for the other `range` just above.

File: src/sage/rings/polynomial/polydict.pyx [hunk 13]
```diff
@@ -754,27 +934,24 @@ cdef class PolyDict:
...
+        D = D.copy()
+        for e, c in R.items():
+            if e in D:
                 D[e] += c
-            except KeyError:
+            else:
                 D[e] = c
-        return PolyDict(D, zero=zero, remove_zero=True,
-                        force_int_exponents=False, force_etuples=False)
+        return self._new(D)
```
Comment (tscrim) [inline src/sage/rings/polynomial/polydict.pyx:1022]: You need to remove zeros here (positive characteristics).

**Review by tscrim — CHANGES_REQUESTED** [review]

File: src/sage/rings/polynomial/multi_polynomial_element.py (hunk 2 above)
Comment (videlec) [reply src/sage/rings/polynomial/multi_polynomial_element.py]: I only changed a space here. I would prefer not touch anything.

Note that the division `polynomial / polynomial` is not complient with the coercion system. To my mind we should just remove `_div_` completely on polynomials and let the coercion framework works for us. But I would prefer postponing this discussion to another issue if that is ok with you. The fraction fields of polynomial rings needs more care than just adjusting this division function.

Comment (tscrim) [reply src/sage/rings/polynomial/multi_polynomial_element.py]: You changed the line just prior and the change produces the exact same output faster (just like your change).

Comment (tscrim) [reply src/sage/rings/polynomial/multi_polynomial_element.py]: Everything happening here is in the base ring.

File: src/sage/rings/polynomial/multi_polynomial_ring.py (hunk 3 above)
Comment (videlec) [reply src/sage/rings/polynomial/multi_polynomial_ring.py]: Why do you want to make a copy here?
```python
sage: R = QQ['x']
sage: S = R['y, z']
sage: S(0) is S(0)
False
```

File: src/sage/rings/polynomial/polydict.pyx (hunk 5 above)
Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx]: What do you refer to with _the argument that disabled this check_?

More importantly the check `type(X) is ETuple` is cythonized as
```
(((PyObject *)Py_TYPE(__pyx_t_1)) == ((PyObject *)__pyx_ptype_4sage_5rings_10polynomial_8polydict_ETuple))
```
You can notice that it converts as a call to `Py_TYPE` (which is an extremely fast inline function) and a single C pointer comparison. In particular, the conversion `ETuple(k)` in the above code is much slower than the type check.

Though, I do not particularly like the `list(v)` which creates a list for possibly no purpose. I will propose an alternative version.

File: src/sage/rings/polynomial/polydict.pyx (hunk 9 above)
Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx:1195]: How do you test `zero`? This is the issue @roed314 complained about: for some base rings (such as power series or p-adics) we do not want to simplify trailing zeros such as `O(t^5)`. Removing zeros is now handled by polynomials in `multi_polynomial_element.py` rather than `PolyDict`. Though I am looking for a better way to consistently handle all base rings (which at the moment is not taken care of at all).

File: src/sage/rings/polynomial/polydict.pyx (hunk 11 above)
Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx:1858]: Interesting. Cython knows well how to optimize `range(a, b)` (for any kind of `a` and `b` which are first converted to the appropriate integral types). However, it does not optimize `range(a, b, c)`.

File: src/sage/rings/polynomial/polydict.pyx (hunk 12 above)
Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx]: Indeed! This is **not** optimized!

File: src/sage/rings/polynomial/multi_polynomial_element.py (hunk 2 above)
Comment (videlec) [reply src/sage/rings/polynomial/multi_polynomial_element.py]: I changed as you suggested.

File: src/sage/rings/polynomial/polydict.pyx (hunk 5 above)
Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx]: See https://github.com/sagemath/sage/pull/35174/commits/3b1b364db624c83e7236e52fa94010e4a542994b

File: src/sage/rings/polynomial/multi_polynomial_ring.py (hunk 3 above)
Comment (tscrim) [reply src/sage/rings/polynomial/multi_polynomial_ring.py]: Speed. Copying the zero object should be faster than going through the coercion framework and (generally) the input checks. (I also am slightly worried about the result not being the same as `S.zero()`; this would be a sensible optimization since elements are supposed to be immutable.)

File: src/sage/rings/polynomial/polydict.pyx (hunk 5 above)
Comment (tscrim) [reply src/sage/rings/polynomial/polydict.pyx]: It still is a cost, whereas before, if you set `force_etuple=False`, then _no_ checks were run. This is code we should really pay attention to the speed aspects.

File: src/sage/rings/polynomial/polydict.pyx (hunk 9 above)
Comment (tscrim) [reply src/sage/rings/polynomial/polydict.pyx:1195]: Note that this is a change in the behavior compared to before. Extra zeros in the polynomial pollute it and make the code slower. Simply do what you do in other places and call the method that you replaced the `remove_zeros` argument.

In fact, I really don't see the utility of pulling that out as a separate method. It leads to easy-to-overlook issues like this and requires more direct manual intervention (with now a seemingly separate method) from the coder. I think you need to have some justification for separating this..

File: src/sage/rings/polynomial/polydict.pyx (hunk 5 above)
Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx]: Can you suggest an example where a speed regression can be noticed?

File: src/sage/rings/polynomial/polydict.pyx (hunk 11 above)
Comment (tscrim) [reply src/sage/rings/polynomial/polydict.pyx:1858]: It is a technical thing with it needing to know the type of `c` is (because of the `i+=c` to `++i` type of things) and they don't really have a good way to do this (even though I think they should really cover some the easier -- or at least trivial -- cases...).

File: src/sage/rings/polynomial/polydict.pyx (hunk 9 above)
Comment (tscrim) [reply src/sage/rings/polynomial/polydict.pyx:1195]: In summary, the here is not how to test for zeros (which is a separate issue), but they fact you are not calling `remove_zeros()` whereas before it was doing the equivalent with `remove_zero=True` for the constructor.

File: src/sage/rings/polynomial/polydict.pyx (hunk 5 above)
Comment (tscrim) [reply src/sage/rings/polynomial/polydict.pyx]: Either a reasonably large polynomial or doing something in a very frequently called loop.

File: src/sage/rings/polynomial/polydict.pyx (hunk 9 above)
Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx:1195]: What do you call polynomial in this context: a `Polydict` or a `MPolynomial_element`?

A `PolyDict` never removes zero coefficients simply because it does not know how to test zero. This test depends on the base ring and `PolyDict` does not store the base ring. However, `MPolynomial_element` does know about the base ring and can implement a `remove_zero`.

File: src/sage/rings/polynomial/polydict.pyx (hunk 5 above)
Comment (tscrim) [reply src/sage/rings/polynomial/polydict.pyx]: Perhaps something like this would be sufficient:
```python
sage: R.<x,y> = PolynomialRing(CyclotomicField(5)['q'].fraction_field())
sage: p = x*y
sage: %timeit p._MPolynomial_element__element.polynomial_coefficient([1,1])
```

File: src/sage/rings/polynomial/polydict.pyx (hunk 9 above)
Comment (tscrim) [reply src/sage/rings/polynomial/polydict.pyx:1195]: As your changes indicate, a `PolyDict` doesn't need to know what the base ring is, it is simply testing where the coefficient thinks it is zero or not. If what you want to say is true about `PolyDict` vs `MPolynomial_element` handling zeros, then there should not be a `PolyDict.remove_zeros()` method. (Looking more closely, if there is a zero, you are actually looping through things twice there. I am not sure if the `all` function with a generator is faster than just manually looping through stuff in Cython,.)

File: src/sage/rings/polynomial/polydict.pyx (hunk 5 above)
Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx]: As a first comparison (data with ~50000 terms)
```python
sage: from sage.rings.polynomial.polydict import ETuple, PolyDict
sage: import itertools
sage: data = {ETuple(t): 1 for t in itertools.product([1r, 2r, 3r, 4r], repeat=8)}
```
gives with the old code
```python
sage: %timeit p = PolyDict(data)
85.3 ms ± 747 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
sage: PolyDict??
sage: %timeit p = PolyDict(data, force_etuples=False)
65.4 ms ± 398 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)
```
while with the new code
```python
sage: %timeit p = PolyDict(data)
1.77 ms ± 9.62 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx]: As a second comparison
```python
sage: R.<x,y> = PolynomialRing(CyclotomicField(5)['q'].fraction_field())
sage: p = x*y
```
gives with the old code
```python
sage: %timeit p._MPolynomial_element__element.polynomial_coefficient([1,1])
2.17 µs ± 18.3 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```
while with the new code
```python
sage: %timeit p._MPolynomial_element__element.polynomial_coefficient([1,1])
1.32 µs ± 9.8 ns per loop (mean ± std. dev. of 7 runs, 1,000,000 loops each)
```

Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx]: Is it sufficient?

File: src/sage/rings/polynomial/polydict.pyx (hunk 9 above)
Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx:1195]: I tend to agree that the current `PolyDict.remove_zeros` is half broken (in term of design, I am not speaking about the code which is certainly not optimal). This is why in the first place this method is only meant to be called by external code that knows what is happening.

I imagine that it would be more appropriate to rather extend it with an optional argument `PolyDict.remove_zeros(zero_test=None)`.

I am very open to any suggestion.

File: src/sage/rings/polynomial/polydict.pyx (hunk 5 above)
Comment (tscrim) [reply src/sage/rings/polynomial/polydict.pyx]: That's not a fair comparison. You have made other optimizations. You would need to implement a parameter for the constructor that skips the `ETuples` checks with the current code.

File: src/sage/rings/polynomial/multi_polynomial_ring.py (hunk 3 above)
Comment (videlec) [reply src/sage/rings/polynomial/multi_polynomial_ring.py]: If elements are supposed immutable, how can you assume that `copy(S.zero())` is not the same object as `S.zero()`? The `copy` just calls `__copy__` on the object which can implement whathever. For example
```python
>>> from copy import copy
>>> a = 1
>>> copy(a) is a
True
```

File: src/sage/rings/polynomial/polydict.pyx (hunk 9 above)
Comment (tscrim) [reply src/sage/rings/polynomial/polydict.pyx:1195]: I think we do want the `PolyDict` to have a way to remove the zeros from itself. The Cython code where it does all of the operations would want to do this and not be predicated on some downstream version in (slow) Python.

Perhaps I am not sure I agree with this (implicit?) separation of a `PolyDict` and a `MPolynomial_element`. It seems like the mathematical model for `PolyDict` really is an element of a polynomial ring, but it is only partially implemented.

I am not quite sure either which is the best way forward. I do see your point about having the actual polynomial control things. Yet, I am worried it now becomes much easier to have cases missed because, e.g., code was working with a `PolyDict`, which used to handle these things automatically...

File: src/sage/rings/polynomial/multi_polynomial_ring.py (hunk 3 above)
Comment (tscrim) [reply src/sage/rings/polynomial/multi_polynomial_ring.py]: I perhaps should have said "or however one implements a copy". Usually there is a faster direct way to make a copy by directly using the class's constructor.

Comment (videlec) [reply src/sage/rings/polynomial/multi_polynomial_ring.py]: Sure, I can do that.

File: src/sage/rings/polynomial/polydict.pyx (hunk 9 above)
Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx:1195]: One way forward could be to have explicit extra argument to operations
```python
class PolyDict:
    def add(self, other, remove_zeros=True, zero_test=None)
    def mul(self, other, remove_zeros=True, zero_test=None)
    ...
```
We can make the default `__add__`, `__mul__`, etc as they used to be (ie remove zero with the standard `bool(coefficient)` for zero testing).

Comment (tscrim) [reply src/sage/rings/polynomial/polydict.pyx:1195]: That would be a bit of an improvement, although I am not sure the extra work and maintenance is worth it from your current proposal.

Could we step back from this local picture and look at what each class is supposed to do? My initial comment might have been looking too locally at the change. I am hoping this will clarify things, possibly leading to your current proposal.

Right now, we have three classes:

1. `ETuple`: represents a single monomial. This is good, unique, and clearly necessary.
2. `PolyDict`: represents a linear combination of terms by a `dict` mapping monomials to their coefficients. Implemented in Cython.
3. `MPolynomial_element`: partially wraps a `PolyDict` but also carries more information about the base ring. Implemented in Python.

Now it is clear what stuff should be handled by the `ETuple`, but it is not clear what separates 2 and 3. Why shouldn't a `PolyDict` know about the coefficient ring? Do we have some reason why it is kept as a separate backend class? Part (most?) of this, I imagine, is to be able to split it off as its own separate entity and not have it fully integrated into Sage. So why doesn't `MPolynomial_element` inherit from `PolyDict`? If it was a Cython class, then we run into multiple inheritance issues. If `MPolynomial_element` inherited from `PolyDict`, then it could implement, e.g.,
```python
def _add_(self, other):
    ret = PolyDict.__add__(self, other)
    ret.remove_zeros()
    return ret
```
which should be comparative in terms of speed (not tested). Although I guess from this separation of the backend with eye towards Cythonization, it would make sense to keep the current version, right?

Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx:1195]: I agree with your analysis of the situation. Some additional points
- making `PolyDict` agnostic to coefficients nature makes it possible to have plain Python type for coefficients (eg `int` or `float`) or mixed types (eg exact and floating point values)
- one extra delicate point with inheritance, is that `PolyDict.__add__` would have to be careful about its return type (which do have a cost)

File: src/sage/rings/polynomial/multi_polynomial_ring.py (hunk 3 above)
Comment (videlec) [reply src/sage/rings/polynomial/multi_polynomial_ring.py]: Hopefully better in https://github.com/sagemath/sage/pull/35174/commits/23d50f4d9a4b6da943dddee3c0ede1bb88f663fd.

Comment (videlec) [reply src/sage/rings/polynomial/multi_polynomial_ring.py]: https://github.com/sagemath/sage/pull/35174/commits/dfd6609d86758031cad774edeb8b078cc2cdeba3 introduces a `check` argument.

File: src/sage/rings/polynomial/polydict.pyx (hunk 5 above)
Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx]: https://github.com/sagemath/sage/pull/35174/commits/dfd6609d86758031cad774edeb8b078cc2cdeba3 introduces a `check` argument.

Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx]: Gets roughly x1.7 faster on the big example above.

File: src/sage/rings/polynomial/polydict.pyx (hunk 9 above)
Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx:1195]: I cleaned the current version to make it clear that we want to use `remove_zeros` more carefully. If you are happy with the current merge request I will open an issue on what has to be done next to handle properly polynomials over rings with trailing zeros.

**Comment by videlec** [comment]
The situation with trailing zeros evaluating to `False` is extremely annoying. For example equality of polynomials is not equality of coefficients
```python
sage: from sage.rings.polynomial.polydict import PolyDict
sage: R.<t> = PowerSeriesRing(QQ)
sage: PolyDict({(1, 0): O(t)}) == PolyDict({})
False
sage: PolyDict({(1, 0): O(t)}) == PolyDict({(1, 0): O(t^2)})
True
```
Similarly, `bool(p)` is probably wrong. Of course the same question holds for matrices, etc.

**Comment by tscrim** [comment]
I will take a look at this tonight.

Indeed, the removal of zeros in general (not just trailing) definitely is useful at keeping this consistent and fast an should be done as much as possible.

File: src/sage/rings/polynomial/polydict.pyx [hunk 14]
```diff
@@ -128,6 +116,9 @@ cdef class PolyDict:
 
         - ``force_etuples`` -- deprecated
 
+        - ``check`` -- if set to ``False`` then assumes that the exponents are all valid ``ETuple``. In
+          that case the construction is a bit faster.
```
Comment (tscrim) [inline src/sage/rings/polynomial/polydict.pyx]:
```suggestion
        - ``check`` -- if set to ``False`` then assumes that the exponents are all valid ``ETuple``; in
          that case the construction is a bit faster
```

File: src/sage/rings/polynomial/polydict.pyx [hunk 15]
```diff
@@ -200,13 +206,53 @@ cdef class PolyDict:
...
+        if not self.__repn:
+            return
+        cdef bint has_zero_coefficient = False
+        if zero_test is None:
+            for coeff in self.__repn.values():
+                if not coeff:
+                    has_zero_coefficient = True
+                    break
+            if not has_zero_coefficient:
+                return
```
Comment (tscrim) [inline src/sage/rings/polynomial/polydict.pyx:247]: Rather than going through everything potentially twice, I would either get a list of keys or do something like
```python
if all(self.__repn.values()):
    return
# otherwise there is at least one 0, so filter it out
```
Similar comment for the `zero_function` case since you had said such generator checks are faster.

File: src/sage/rings/polynomial/polydict.pyx [hunk 16]
```diff
@@ -2305,68 +2387,92 @@ cdef class ETuple:
...
+        result._data = <int*> sig_malloc(sizeof(int) * 2 * self._nonzero)
+        exp1 = self._data[2 * ind + 1]
+        if exp1 > 1:
+            # division doesn't change the number of nonzero positions
+            result._nonzero = self._nonzero
+            memcpy(result._data, self._data, sizeof(int) * 2 * self._nonzero)
+            result._data[2 * ind + 1] = exp1 - 1
+        else:
+            # var(pos) disappears from self
+            result._nonzero = self._nonzero-1
```
Comment (tscrim) [inline src/sage/rings/polynomial/polydict.pyx]:
```suggestion
            result._nonzero = self._nonzero - 1
```
PEP8

**Review by tscrim — CHANGES_REQUESTED** [review]

File: src/sage/rings/polynomial/polydict.pyx (hunk 9 above)
Comment (tscrim) [reply src/sage/rings/polynomial/polydict.pyx:1195]: I think I am modulo the last few things I commented below about with your more recent changes. It is nice to see a little more clarification in the docs about this.

File: src/sage/rings/polynomial/multi_polynomial_ring.py (hunk 3 above)
Comment (tscrim) [reply src/sage/rings/polynomial/multi_polynomial_ring.py]: I think that is good. Thank you.

File: src/sage/rings/polynomial/polydict.pyx (hunk 9 above)
Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx:1195]: I opened a general discussion about inexact zeros at #35319. I will mention the latter in the documentations of `PolyDict` and `MPolynomial_element`.

File: src/sage/rings/polynomial/polydict.pyx (hunk 15 above)
Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx:247]: Actually `all` is very badly handled by `cython`. The loop at 220-225 is precisely unfolding this `all`.

Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx:247]: Comment added in https://github.com/sagemath/sage/pull/35174/commits/1a7eeac1a5da609b30fa9f0a36d6784e5dff6e7d

File: src/sage/rings/polynomial/polydict.pyx (hunk 14 above)
Comment (videlec) [reply src/sage/rings/polynomial/polydict.pyx]: Fixed in https://github.com/sagemath/sage/pull/35174/commits/1a7eeac1a5da609b30fa9f0a36d6784e5dff6e7d

**Review by tscrim — APPROVED** [review]
Thank you. LGTM. We can always do more changes later should we find other improvements.
